### DISCUSSION TITLE
How to get the velocity of the ego vehicle?

### DISCUSSION THREAD

**Discussion by vikram0230** [carla-simulator/carla#4947] [Q&A]
How to get the speed of the vehicle running on auto-pilot? I have checked the carla docs but couldn't figure it out. I found that there is a workaround, by calculating the velocity by location of the vehicle relative to time. 
Is there a better way to get the velocity of the ego?

**Comment by varunjammula** [comment]
You can use get_velocity() [https://carla.readthedocs.io/en/latest/python_api/](https://carla.readthedocs.io/en/latest/python_api/)

**Comment by sbv198** [comment]
@varunjammula i used the same method to get the velocity.

It gives 3D vector with x,y,z cordinates.  After that did you use 3.6*sqrt(x**2 +y**2+z**2)?

for my project i need longitudinal and lateral velocity 

can you help out in further detail?

**Comment by SExpert12** [comment]
Have you got the solution for your answer? I also want to know longitudinal and lateral velocity of vehicle for my project.

**Reply by sbv198** [reply]
Yess you can use  get_velocity(). That will give you velocity in x,y,z. Velocity in y is lateral velocity to get directly use get_velocity().y

**Comment by SExpert12** [comment]
Okay. and with this equation 3.6*sqrt(x2 +y2+z**2), it gives me longitudinal velocity?

**Reply by sbv198** [reply]
It will be norm velocity. But be careful with it. It depends on a cartesian coordinates so if your vehicle is positioned let’s say (x=100, y=200) and you are moving in y direction with some velocity. get_velocity() will give you  x=100 with some deviation and y will be changing so do not consider y = lateral velocity! 

Your vehicle coordinates changes with direction of travel while cartesian stays same ! 

For above mentioned example 

x=y and y =x for Cartesian to vehicle coordinates

**Comment by SExpert12** [comment]
Okay. I understood the concept. And yes you have clarified the concept that velocity depends on the travel direction. Let's say I am moving my vehicle in four direction e.g  north, south, east and west. Is there any way to fix my longitudinal velocity and lateral velocity in this case. As carla is giving me 3D vector, it's really confusing. I want to give specific lateral velocity and longitudinal velocity based on specific case.

Thanks once again for sparing time to share your knowledge.

**Reply by 9900ff** [reply]
Hi, would you please share with us what you have ended up doing eventually to get the longitudinal and lateral velocities?

**Comment by Anhar2** [comment]
> Okay. I understood the concept. And yes you have clarified the concept that velocity depends on the travel direction. Let's say I am moving my vehicle in four direction e.g north, south, east and west. Is there any way to fix my longitudinal velocity and lateral velocity in this case. As carla is giving me 3D vector, it's really confusing. I want to give specific lateral velocity and longitudinal velocity based on specific case.
> 
> Thanks once again for sparing time to share your knowledge.


Hi, would you please share with us what you have ended up doing eventually to get the longitudinal and lateral velocities?

**Comment by SExpert12** [comment]
I want to control vehicle in the intersection area.
